src/api/exchange/ExchangeApi.ts

```diff
@@ -101,21 +115,175 @@ export interface Order {
   remainingAmount: BN
 }
 
+/**
+ * BaseTradeEvent uses only info available on the emitted Trade event
+ */
+export interface BaseTradeEvent {
+  // order related
+  orderId: string
+  sellTokenId: number
+  buyTokenId: number
+  sellAmount: BN
+  buyAmount: BN
+  // block related
+  txHash: string
+  txIndex: number
+  blockNumber: number
+  id: string // txHash + | + txIndex
+}
+
+/**
+ * TradeEvent enriches BaseTradeEvent with block, order and token data
+ */
+export interface TradeEvent extends BaseTradeEvent {
+  batchId: number
+  hashKey: string // orderId + batchId, to find reverts
+  indexOnBatch: number // tracks trade position on batch, in case of reverts
+  time: Date
+  buyToken: TokenDetails
+  sellToken: TokenDetails
+  limitPrice: BigNumber
+  fillPrice: BigNumber
+  reverted: boolean
+}
+
 export interface GetOrdersPaginatedResult {
   orders: AuctionElement[]
   nextIndex?: number
 }
 
+// TODO: move to const/config
+const CONTRACT_DEPLOYMENT_BLOCK = {
+  1: 9340147,
+  4: 5844678,
+}
+
+// Syntactic sugar. Aliasing event types to make code cleaner
+type TradeSubscription = ReturnType<BatchExchangeContract['events']['Trade']>
```

Comment:
I like it 👌

Although I think in one PR I suggested something like this, and @Velenir commented that this could bring problems. I think it was with the flattening of the Promises, so it might be different case  🤷‍♂️